### PR TITLE
[eslint config] [base] [breaking] Add no-plusplus in base errors.js and desc in README

### PR DESCRIPTION
**Pull request by sstern6**
@ljharb  Added `no-plusplus: 'error'` to eslint-config-airbnb-base/errors.js in order to disallow unary increment and decrement statements.  In addition, added description under Variables in the main README. 

This PR references https://github.com/airbnb/javascript/issues/540.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1418,6 +1418,35 @@ Other Style Guides
     }
     ```
 
+    <a name="variables--unary-increment-decrement"></a><a name="13.5"></a>
```
Comment (ljharb) [inline README.md]: there's no need to add a link to "13.5" - the numbers only exist for legacy reasons.

File: README.md [hunk 2]
```diff
@@ -1418,6 +1418,35 @@ Other Style Guides
...
+    - [13.5](#variables--unary-increment-decrement) Avoid using unary increments and decrements (++, --). Eslint [`no-plusplus`](http://eslint.org/docs/rules/no-plusplus)
```
Comment (ljharb) [inline README.md]: `eslint` should be lowercased

File: README.md [hunk 3]
```diff
@@ -1418,6 +1418,35 @@ Other Style Guides
...
+
+      > Why? Per the Eslint documentation, unary increment and decrement statements are subject to automatic semicolon insertion and can cause silent errors with incrementing or decrementing values within an application.  It is also more expressive to mutate your values with statements like `num += 1` instead of `num ++`.  Disallowing unary increment and decrement statements also prevents you from pre-incrementing/pre-decrementing values unintentionally which can also cause unexpected behavior in your programs.
```
Comment (ljharb) [inline README.md]: `eslint` should be lowercased

File: README.md [hunk 4]
```diff
@@ -1418,6 +1418,35 @@ Other Style Guides
...
+        }
+
+        // good
+
+        let array = [1, 2, 3];
+        let num = 1;
+        let increment = num += 1;
+        let decrement = num -= 1;
+
+        for(let i = 0; i < array.length; i += 1){
```
Comment (ljharb) [inline README.md]: in this case, the guide discourages ever using loops, so the "good" example should be
```
array.forEach((item) => {
  doSomethingWithItem(item);
});
```

File: packages/eslint-config-airbnb-base/rules/errors.js [hunk 5]
```diff
@@ -74,6 +74,10 @@ module.exports = {
     // http://eslint.org/docs/rules/no-prototype-builtins
     'no-prototype-builtins': 2,
 
+    // disallow use of unary operators, ++ and --
+    // http://eslint.org/docs/rules/no-plusplus
+    'no-plusplus': 'error',
```
Comment (ljharb) [inline packages/eslint-config-airbnb-base/rules/errors.js]: please use `2` here to remain consistent with the rest of the file

Comment (ljharb) [inline packages/eslint-config-airbnb-base/rules/errors.js]: Also, according to http://eslint.org/docs/rules/#stylistic-issues this rule is in "style", so please delete the relevant lines from this file, and move them to lines 183-185 in `rules/style.js`

File: README.md [hunk 6]
```diff
@@ -1418,6 +1418,38 @@ Other Style Guides
...
+
+      ```javascript
+        // bad
+
+        let array = [1, 2, 3];
+        let num = 1;
+        let increment = num ++;
+        let decrement = -- num;
+
+        array.forEach((value) => {
```
Comment (ljharb) [inline README.md]: sorry if i was unclear before; i think having a loop in the "bad" example is actually a good thing :-)

Comment (sstern6) [inline README.md]: Ah ok, I apologize, I misunderstood. I interpreted it incorrectly.  Fixing now.

**Comment by ljharb** [comment]
Thanks! I'd prefer this rebased down to a single commit, if you don't mind.

I'm also going to wait a while before merging this, since it's a breaking change.

**Comment by sstern6** [comment]
Sure no problem, will have the rebase done tonight!  Thanks for the feedback through this PR

**Comment by tleunen** [comment]
What do you think of allowing it in loops?

Sometimes, a basic for loop is used in order to do things and break/return something without doing through the entire array. And usually people are using `i++`, not really `i += 1`.

**Comment by ljharb** [comment]
@tleunen if there was a way for the linter rule to allow it in loops, i'd consider that, but since this guide forbids any and all loops at any time for any reason, I'm not sure why someone following the guide would even run into the problem? (There's `Array#some`, `Array#every`, and `Array#find` for "stop partway through" logic)

**Comment by sstern6** [comment]
@ljharb just for reference, there is an arg you can send in to no-plusplus that addresses the for loop issue.` "allowForLoopAfterthoughts": true `.  I understand that the guides forbid the use of  loops.  Just pointing it out for reference.

**Comment by ljharb** [comment]
I'll think about whether we should add that, since there's not really as much of a need to avoid unary increment/decrement in for loop position.

File: README.md [hunk 7]
```diff
@@ -1418,6 +1418,35 @@ Other Style Guides
     }
     ```
 
+    - [13.5](#variables--unary-increment-decrement) Avoid using unary increments and decrements (++, --). eslint [`no-plusplus`](http://eslint.org/docs/rules/no-plusplus)
+
+      > Why? Per the eslint documentation, unary increment and decrement statements are subject to automatic semicolon insertion and can cause silent errors with incrementing or decrementing values within an application.  It is also more expressive to mutate your values with statements like `num += 1` instead of `num ++`.  Disallowing unary increment and decrement statements also prevents you from pre-incrementing/pre-decrementing values unintentionally which can also cause unexpected behavior in your programs.
```
Comment (ljharb) [inline README.md]: sorry to be picky, but outside of typewriters there's no need for two spaces after a period :-)

File: packages/eslint-config-airbnb/package.json [hunk 8]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "eslint-config-airbnb",
-  "version": "10.0.0",
+  "version": "10.0.1",
```
Comment (ljharb) [inline packages/eslint-config-airbnb/package.json]: i think you may need to rebase instead of merge?

File: README.md (hunk 7 above)
Comment (sstern6) [inline README.md]: No worries, thanks for the feedback, will fix.

File: packages/eslint-config-airbnb/package.json (hunk 8 above)
Comment (sstern6) [inline packages/eslint-config-airbnb/package.json]: Yep, rebased and repushed. Should be good now

**Comment by ljharb** [comment]
@sstern6 hm, why does this commit have me listed as the author?

**Comment by sstern6** [comment]
Ive never seen that before, this morning there were merge conflicts, I:

1) Pulled origin
2) Rebased against master
3) Pushed

It could have reassigned authorship when I rebased?  I will look into possible reasons and solutions.  Do you have any ideas?

**Comment by ljharb** [comment]
Simplest might be to soft reset back to master, and make a new commit (on the same branch), and then force push that up?

I can think of any number of very unlikely ways this would happen, but I have no idea how this one would have, unless you started by cherry-picking one of my commits and amended it?

**Comment by sstern6** [comment]
Ya sounds good!  No I did not do that.  Im also going to look at the reflogs and see if I can find where it went wrong for future reference.  I will reset soft back to master and make a new commit with the final changes then push!

Thanks

**Comment by sstern6** [comment]
@ljharb just force pushed after resetting the commit, looked like in the rebase interactive, first squashing the commits amended your commit which is why I think you were granted authorship.  Everything should be up to date, let me know if you have any feedback or recommend any additional changes.

Thanks

**Comment by sstern6** [comment]
@ljharb with this being a breaking change, how long do you wait until merging and what does that process look like, "merging an update that contains a breaking change in a large codebase".  1) for curiosity and 2) in order to time my rebase to make sure were green when you're ready to merge.

Thanks

**Comment by ljharb** [comment]
@sstern6 totally fair question - generally i'd want to wait until we have a few major changes queued up for the base config, and then merge them all at once.

It might be a couple weeks - if you don't mind waiting, I'll ping you when the time is right :-)

**Comment by sstern6** [comment]
Perfect thank you for the clarification.

**Comment by ljharb** [comment]
@sstern6 do you mind rebasing? I'll merge and release this now.

**Comment by sstern6** [comment]
REBASING NOW, STANDBY!

**Comment by sstern6** [comment]
@ljharb should be solid, if anything needs to be adjusted please let me know.

Thanks

**Pull request merged**